daemon/containerd/image_import.go

```diff
@@ -258,7 +258,6 @@ func compressAndWriteBlob(ctx context.Context, cs content.Store, compression arc
 	if err != nil {
 		return "", "", errdefs.InvalidParameter(err)
 	}
-	defer compressor.Close()
```

Comment:
Is there any situation where `archive.CompressStream` _should_ be closed? (mostly thinking if it should be returning a `io.Nopcloser` in that case)

If it's specific to _this_ case, perhaps we should replace this line with a comment to outline  "why".

Also wondering if something like a [io.TeeReader](https://pkg.go.dev/io#TeeReader) would be something to involve 🤔

Reply:
It is closed later down

Reply:
In that case, would the _reverse_ work? Keep this `defer`, and remove the other close further down? (would that be more readable?) The defers are executed in reverse order, so in this case,

```go
defer pr.Close()
defer pw.Close()
defer compressor.Close()
```

Would be executed as;

```go
pw.CloseWithError(err)
pr.CloseWithError(err)

# defers:
compressor.Close()
pw.Close()  // redundant, but ok
pr.Close()  // redundant, but ok
```

Reply:
No, the compressor must be closed before `compressedDigest := <-writeChan` and before `pr` and `pw` are closed. 

Closing `pr` or `pw` before `compressor` closes the writer the compressor writes to, so it doesn't have the chance to write all data.

Reply:
DOH! Looking at that channel twice, and still not considering it for the order of events in my comment. 🙈 yup, makes sense.